pyomo/core/base/constraint.py

```diff
@@ -331,6 +331,58 @@ def __init__(self,  expr=None, component=None):
         if expr is not None:
             self.set_value(expr)
 
+        self._canonical_form_has_been_set = False
+        self._canonical_form = None
+
+    @property
+    def _linear_canonical_form(self):
```

Comment:
Used by `Pyomo` code already to check whether we expect the object to have a `canonical_form()` method. This is because `linear_constraint` from the `kernel` library provides this interface.